Associated constants in trait functions

This is a weird one & may not be an issue with `async-trait`, so please close if needed.  When trying to use associated constants in macros (such as `debug!` or `println!`) it appears to trip up the compiler.

This standard trait compiles:

```rust
pub trait Example {
    const ASSOCIATED: &'static str;

    fn associated(&self) {
        println!("Associated:{}", Self::ASSOCIATED);
    }
}
```

This decorated trait does not compile:

```rust
#[async_trait]
pub trait Example {
    const ASSOCIATED: &'static str;

    async fn associated(&self) {
        println!("Associated:{}", Self::ASSOCIATED);
    }
}
```

I get the following error message:
```
error[E0401]: can't use generic parameters from outer function
  --> src/example.rs:16:35
   |
16 |         println!("Associated:{}", Self::ASSOCIATED);
   |                                   ^^^^^^^^^^^^^^^^
   |                                   |
   |                                   use of generic parameter from outer function
   |                                   can't use `Self` here
```

However, the trait impl works fine if you `let` bind it:

```rust
#[async_trait]
pub trait Example {
    const ASSOCIATED: &'static str;

    async fn associated(&self) {
        let associated = Self::ASSOCIATED;
        println!("Associated:{}", associated);
    }
}
```

The workaround appears to be OK for my use case.

--- a/src/receiver.rs
+++ b/src/receiver.rs
@@ -205,6 +205,7 @@ fn fold_token_stream(tokens: &mut TokenStream) -> bool {
         match tt {
             TokenTree::Ident(mut ident) => {
                 modified |= prepend_underscore_to_self(&mut ident);
+                modified |= replace_self_from_outer_function(&mut ident);
                 out.push(TokenTree::Ident(ident));
             }
             TokenTree::Group(group) => {
@@ -227,6 +228,14 @@ fn prepend_underscore_to_self(ident: &mut Ident) -> bool {
     let modified = ident == "self";
     if modified {
         *ident = Ident::new("_self", ident.span());
+    }
+    modified
+}
+
+fn replace_self_from_outer_function(ident: &mut Ident) -> bool {
+    let modified = ident == "Self";
+    if modified {
+        *ident = Ident::new("AsyncTrait", ident.span());
     }
     modified
 }

--- a/tests/test.rs
+++ b/tests/test.rs
@@ -502,3 +502,17 @@ pub mod issue57 {
         executor::block_on_simple(fut);
     }
 }
+
+// https://github.com/dtolnay/async-trait/issues/73
+pub mod issue73 {
+    use async_trait::async_trait;
+
+    #[async_trait]
+    pub trait Example {
+        const ASSOCIATED: &'static str;
+
+        async fn associated(&self) {
+            println!("Associated:{}", Self::ASSOCIATED);
+        }
+    }
+}

